### PR TITLE
Fix material switching for objects using shaders (ogre2)

### PR DESCRIPTION
**Pull request by iche033**
# 🦟 Bug fix

Related PR: https://github.com/ignitionrobotics/ign-gazebo/pull/1275

## Summary

Mouse picking works by assigning a unique material to an object. When a user clicks on a object, we switch out its material to a unique one to identify the clicked object. There is bug that when switching objects that use custom shader materials. Added an extra check and logic make sure we restore the correct material (low level shader material).

Also fixed a bug that causes ign-rendering to crash when a long shader name is given.

Test this with ign-gazebo using https://github.com/ignitionrobotics/ign-gazebo/pull/1275. Launch the `shader_param.sdf` world and hover the mouse over the 3D render window. Without these changes, the shader will stop working and the sphere will turn black.

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Review by ahcorde — CHANGES_REQUESTED** [review]
can we add a test ?

**Comment by iche033** [comment]
> can we add a test ?

yep, added test in 33dd4c4

**Comment by iche033** [comment]
hmm test is failing on github bionic build but pass on focal and other jenkins machines. Not sure if it's a bionic software rendering or mesa driver issue or something else. Added logic to disable test on github action. f82a0ac

File: test/integration/camera.cc [hunk 1]
```diff
@@ -552,6 +561,165 @@ void CameraTest::Visibility(const std::string &_renderEngine)
   rendering::unloadEngine(engine->Name());
 }
 
+/////////////////////////////////////////////////
+void CameraTest::ShaderSelection(const std::string &_renderEngine)
+{
+  if (_renderEngine == "optix")
+  {
+    igndbg << "Custom shaders are not supported yet in rendering engine: "
+            << _renderEngine << std::endl;
```
Comment (ahcorde) [inline test/integration/camera.cc:570]:
```suggestion
           << _renderEngine << std::endl;
```

File: test/integration/camera.cc [hunk 2]
```diff
@@ -552,6 +561,165 @@ void CameraTest::Visibility(const std::string &_renderEngine)
...
+  // This test checks that custom shaders are being rendering correctly in
+  // camera view. It also verifies that visual selection is working and the
+  // visual's material remains the same after selection.
+
+  // create and populate scene
+  RenderEngine *engine = rendering::engine(_renderEngine);
+  if (!engine)
+  {
+    igndbg << "Engine '" << _renderEngine
+              << "' is not supported" << std::endl;
```
Comment (ahcorde) [inline test/integration/camera.cc:600]:
```suggestion
    igndbg << "Engine '" << _renderEngine
           << "' is not supported" << std::endl;
```

**Review by ahcorde — APPROVED** [review]
some minor style issues

File: test/integration/camera.cc (hunk 2 above)
Comment (iche033) [reply test/integration/camera.cc]: dc5211e17560836ddc900ea4fabb8047d55ad6f9

File: test/integration/camera.cc (hunk 1 above)
Comment (iche033) [reply test/integration/camera.cc]: dc5211e17560836ddc900ea4fabb8047d55ad6f9

**Review by ahcorde — APPROVED** [review]

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-03-01-citadel-edifice-fortress/1313/1

**Pull request merged**